### ISSUE TITLE
Make it easier to publish shaded library jars

### ISSUE THREAD

**Issue by jrudolph** [sbt/sbt-assembly#365]
https://github.com/sbt/sbt-assembly#q-despite-the-concerned-friends-i-still-want-publish-fat-jars-what-advice-do-you-have

explains how can publish a fat-jar. That's not the only use-case for publishing an assembled jar. Another use case, is that your library depends on a third-party lib it doesn't want to depend directly on. Instead, you want to publish a module/artifact with a shaded version. As shown in the README paragraph above that currently needs an extra sub-module. The reason is that you can otherwise easily create cyclic dependencies in the sbt task graph when you try to use something like this:

```scala
    libraryDependencies += Dependencies.Compile.jctools, // want to shade this
    assemblyShadeRules in assembly := Seq(
      ShadeRule.rename("org.jctools.**" -> "akka.jctools.@1").inLibrary(Dependencies.Compile.jctools)
    ),
    assemblyOption in assembly := (assemblyOption in assembly).value.copy(includeScala = false, includeBin = false), // only need the dependency and nothing else
    exportJars := true,
    packageBin in Compile := (assembly in Compile).value
```

This will silently get stuck when running `assembly` or `packageBin` because `exportJars := true` introduce a dynamic task dependency from `exportedProduct` -> `assembly`. Assembly on the other hand depends on `packageBin` because of 1) delegating to `fullClasspath` and 2) running `test` (for no good reason).

To break the tie, I can add those settings to make it work:

```scala
    fullClasspath in assembly := (externalDependencyClasspath in Compile).value, // otherwise, there's a cyclic dependency between packageBin and assembly
    test in assembly := {}, // assembly runs tests for unknown reason which introduces another cyclic dependency to packageBin via exportedJars
```

Would be nice, if that use case would be supported out of the box or at least be documented.

**Comment by eed3si9n** [comment]
Shaded JAR cannot be used as a library, unless it's pure Java.
Jarjar does not migrate the Scala pickles, so the compilation would fail.

**Comment by jrudolph** [comment]
> Shaded JAR cannot be used as a library, unless it's pure Java.

That's true but in my case it's pure Java. That shading doesn't work for Scala libraries is basically another topic.
